infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -517,6 +562,54 @@ protected String evalPlaceholders(String namespace, NamespaceResolutionContext c
     return evaluated;
   }
 
+  /**
+   * Stores computed namespace name and it's template checksum into user preferences. Checksum is
+   * required to track template changes and re-generate namespace in case it didn't matches.
+   */
+  private void recordEvaluatedNamespaceName(String namespace) {
+    try {
+      String owner = EnvironmentContext.getCurrent().getSubject().getUserId();
+      Map<String, String> preferences = preferenceManager.find(owner);
+      preferences.put(WORKSPACE_INFRASTRUCTURE_NAMESPACE_ATTRIBUTE, namespace);
+      preferences.put(NAMESPACE_TEMPLATE_CHECKSUM_ATTRIBUTE, crc32(defaultNamespaceName));
+      preferenceManager.update(owner, preferences);
+    } catch (ServerException e) {
+      LOG.error(e.getMessage(), e);
+    }
+  }
+
+  /** Returns stored namespace if any, and its default template checksum. */
+  private Optional<Pair<String, String>> getPreferencesNamespaceName(
+      NamespaceResolutionContext context) {
+    try {
+      String owner = context.getUserId();
+      Map<String, String> preferences = preferenceManager.find(owner);
+      if (preferences.containsKey(WORKSPACE_INFRASTRUCTURE_NAMESPACE_ATTRIBUTE)
+          && preferences.containsKey(NAMESPACE_TEMPLATE_CHECKSUM_ATTRIBUTE)) {
+        return Optional.of(
+            Pair.of(
+                preferences.get(WORKSPACE_INFRASTRUCTURE_NAMESPACE_ATTRIBUTE),
+                preferences.get(NAMESPACE_TEMPLATE_CHECKSUM_ATTRIBUTE)));
+      }
+    } catch (ServerException e) {
+      LOG.error(e.getMessage(), e);
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Checks that current template does not contains workspace id and compares current and stored
+   * template checksum.
+   */
+  private boolean isStoredTemplateValid(String storedNamespaceChecksum) {
+    if (defaultNamespaceName.contains(WORKSPACEID_PLACEHOLDER)) {
```

Comment:
do we need to store the namespace if it does not have any placeholder? a.k.a. single namespace for all workspaces strategy.

Reply:
Imo for consistency is better to always save if `che.infra.kubernetes.namespace.allow_user_defined` is `true`